File: packages/flutter-dart/lib/appkit_smart_account/smart_account_manager.dart
import 'package:flutter/foundation.dart';
import 'user_operation.dart';
import 'bundler_client.dart';

/// ERC-4337 Smart Account Manager
class SmartAccountManager extends ChangeNotifier {
  final String bundlerUrl;
  static const _entryPointAddress = '0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789';

  SmartAccountManager({
    this.bundlerUrl = 'https://bundler.cinacoin.com',
  });

  String? _smartAccountAddress;
  String? get smartAccountAddress => _smartAccountAddress;

  bool _isDeployed = false;
  bool get isDeployed => _isDeployed;

  Future<String> getSmartAccount({
    required String ownerAddress,
    int salt = 0,
  }) async {
    final address = _computeAddress(ownerAddress, salt);
    _smartAccountAddress = address;
    _isDeployed = await _checkDeployed(address);
    notifyListeners();
    return address;
  }

  Future<String> deploy({
    required String ownerAddress,
    int salt = 0,
  }) async {
    final address = _smartAccountAddress ?? _computeAddress(ownerAddress, salt);
    final initCode = _buildInitCode(ownerAddress, salt);

    final userOp = UserOperation(
      sender: address,
      nonce: 0,
      initCode: initCode,
      callData: [],
      callGasLimit: 500000,
      verificationGasLimit: 500000,
      preVerificationGas: 50000,
      maxFeePerGas: 1000000000,
      maxPriorityFeePerGas: 1000000000,
      paymasterAndData: [],
      signature: [],
    );

    final txHash = await BundlerClient.sendUserOperation(
      userOp: userOp,
      entryPoint: _entryPointAddress,
      bundlerUrl: bundlerUrl,
    );

    _isDeployed = true;
    notifyListeners();
    return txHash;
  }

  Future<String> execute({
    required String target,
    required int value,
    required List<int> data,
    required String ownerAddress,
  }) async {
    final accountAddress = _smartAccountAddress;
    if (accountAddress == null) throw const NotInitializedError();

    final callData = _encodeExecute(target, value, data);
    final nonce = await _getNonce(accountAddress);

    final userOp = UserOperation(
      sender: accountAddress,
      nonce: nonce,
      initCode: [],
      callData: callData,
      callGasLimit: 500000,
      verificationGasLimit: 500000,
      preVerificationGas: 50000,
      maxFeePerGas: 1000000000,
      maxPriorityFeePerGas: 1000000000,
      paymasterAndData: [],
      signature: [],
    );

    return BundlerClient.sendUserOperation(
      userOp: userOp,
      entryPoint: _entryPointAddress,
      bundlerUrl: bundlerUrl,
    );
  }

  // Placeholder implementations
  String _computeAddress(String owner, int salt) => '0x${'0' * 40}';
  Future<bool> _checkDeployed(String address) async => false;
  List<int> _buildInitCode(String owner, int salt) => [];
  List<int> _encodeExecute(String target, int value, List<int> data) => [];
  Future<int> _getNonce(String address) async => 0;
}

File: packages/ios-swift/Sources/CinacoinAppKitSmartAccount/SmartAccountManager.swift
import Foundation
import CinacoinAppKitConfig

/// ERC-4337 Smart Account Manager
public final class SmartAccountManager: ObservableObject, @unchecked Sendable {
    
    @Published public private(set) var smartAccountAddress: String?
    @Published public private(set) var isDeployed = false
    
    private let bundlerUrl: String
    private let entryPointAddress = "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789"
    
    public init(bundlerUrl: String = "https://bundler.cinacoin.com") {
        self.bundlerUrl = bundlerUrl
    }
    
    /// Create or get existing smart account
    public func getSmartAccount(ownerAddress: String, salt: UInt64 = 0) async throws -> String {
        // Compute counterfactual address
        let address = try await computeAddress(owner: ownerAddress, salt: salt)
        smartAccountAddress = address
        
        // Check if deployed
        isDeployed = try await checkDeployed(address: address)
        
        return address
    }
    
    /// Deploy the smart account
    public func deploy(ownerAddress: String, salt: UInt64 = 0) async throws -> String {
        let initCode = try buildInitCode(owner: ownerAddress, salt: salt)
        
        let userOp = UserOperation(
            sender: try await computeAddress(owner: ownerAddress, salt: salt),
            nonce: 0,
            initCode: initCode,
            callData: Data(),
            callGasLimit: 500_000,
            verificationGasLimit: 500_000,
            preVerificationGas: 50_000,
            maxFeePerGas: 1_000_000_000,
            maxPriorityFeePerGas: 1_000_000_000,
            paymasterAndData: Data(),
            signature: Data()
        )
        
        let txHash = try await BundlerClient.sendUserOperation(
            userOp: userOp,
            entryPoint: entryPointAddress,
            bundlerUrl: bundlerUrl
        )
        
        isDeployed = true
        return txHash
    }
    
    /// Execute a transaction via the smart account
    public func execute(
        target: String,
        value: UInt256,
        data: Data,
        ownerAddress: String
    ) async throws -> String {
        guard let accountAddress = smartAccountAddress else {
            throw SmartAccountError.notInitialized
        }
        
        let callData = try encodeExecute(target: target, value: value, data: data)
        
        let userOp = UserOperation(
            sender: accountAddress,
            nonce: try await getNonce(address: accountAddress),
            initCode: Data(),
            callData: callData,
            callGasLimit: 500_000,
            verificationGasLimit: 500_000,
            preVerificationGas: 50_000,
            maxFeePerGas: 1_000_000_000,
            maxPriorityFeePerGas: 1_000_000_000,
            paymasterAndData: Data(),
            signature: Data() // To be signed by owner
        )
        
        return try await BundlerClient.sendUserOperation(
            userOp: userOp,
            entryPoint: entryPointAddress,
            bundlerUrl: bundlerUrl
        )
    }
    
    /// Batch execute multiple transactions
    public func executeBatch(
        targets: [String],
        values: [UInt256],
        datas: [Data],
        ownerAddress: String
    ) async throws -> String {
        guard targets.count == values.count, values.count == datas.count else {
            throw SmartAccountError.invalidBatch
        }
        
        let callData = try encodeExecuteBatch(targets: targets, values: values, datas: datas)
        
        guard let accountAddress = smartAccountAddress else {
            throw SmartAccountError.notInitialized
        }
        
        let userOp = UserOperation(
            sender: accountAddress,
            nonce: try await getNonce(address: accountAddress),
            initCode: Data(),
            callData: callData,
            callGasLimit: UInt64(500_000 * targets.count),
            verificationGasLimit: 500_000,
            preVerificationGas: 50_000,
            maxFeePerGas: 1_000_000_000,
            maxPriorityFeePerGas: 1_000_000_000,
            paymasterAndData: Data(),
            signature: Data()
        )
        
        return try await BundlerClient.sendUserOperation(
            userOp: userOp,
            entryPoint: entryPointAddress,
            bundlerUrl: bundlerUrl
        )
    }
    
    // MARK: - Private
    
    private func computeAddress(owner: String, salt: UInt64) async throws -> String {
        // In production: compute CREATE2 address from factory + init code
        return "0x\(String(repeating: "0", count: 40))" // Placeholder
    }
    
    private func checkDeployed(address: String) async throws -> Bool {
        // Check code at address via RPC
        return false // Placeholder
    }
    
    private func buildInitCode(owner: String, salt: UInt64) throws -> Data {
        // Factory address + createAccount call
        return Data() // Placeholder
    }
    
    private func encodeExecute(target: String, value: UInt256, data: Data) throws -> Data {
        // encode function execute(address dest, uint256 value, bytes func)
        return Data() // Placeholder
    }
    
    private func encodeExecuteBatch(targets: [String], values: [UInt256], datas: [Data]) throws -> Data {
        // encode function executeBatch(address[] dest, uint256[] value, bytes[] func)
        return Data() // Placeholder
    }
    
    private func getNonce(address: String) async throws -> UInt64 {
        // Call entryPoint.getNonce(address, 0)
        return 0 // Placeholder
    }
}

/// Smart account errors
public enum SmartAccountError: Error, LocalizedError {
    case notInitialized
    case invalidBatch
    case deploymentFailed
    case userOpFailed(String)
    
    public var errorDescription: String? {
        switch self {
        case .notInitialized: return "Smart account not initialized"
        case .invalidBatch: return "Invalid batch: arrays must have same length"
        case .deploymentFailed: return "Smart account deployment failed"
        case .userOpFailed(let msg): return "User operation failed: \(msg)"
        }
    }
}

/// Placeholder for UInt256
public typealias UInt256 = UInt64
